### PR TITLE
Pipeline/Bin subclassing

### PR DESCRIPTION
**Pull request by frostbyte73**
Moves all linking, state change, and EOS code into the pipeline/bin classes

### CODE REVIEW

File: pkg/gstreamer/bin.go [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+	"sync"
+
+	"github.com/tinyzimmer/go-gst/gst"
+
+	"github.com/livekit/egress/pkg/errors"
+	"github.com/livekit/protocol/logger"
+)
+
+// Bins are designed to hold a single stream, with any number of sources and sinks
+type Bin struct {
```
Comment (davidzhao) [inline pkg/gstreamer/bin.go:28]: Looks like an useful abstraction! Some comments about general usage would be helpful for me personally. looks like this has potential of being in a livekit-gst package? (shared between our multiple apps)

**Review by davidzhao — COMMENTED** [review]
awesome work! just a few comments.

File: pkg/gstreamer/bin.go (hunk 1 above)
Comment (frostbyte73) [reply pkg/gstreamer/bin.go:28]: I did have that in mind when writing it, aiming do that after PC. Right now it has a couple egress-specific things like callbacks, and it needs more work to make sure (or at least explain how) it's used properly. Can definitely add more comments for now

File: pkg/gstreamer/pipeline.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+func (p *Pipeline) SendEOS() {
+	p.sendEOS()
+}
+
+func (p *Pipeline) Stop() {
+	if err := p.SetState(gst.StateNull); err != nil {
+		p.OnError(err)
+		return
+	}
+	if err := p.OnStop(); err != nil {
```
[the remark this answers is not in the document]
Comment (frostbyte73) [reply pkg/gstreamer/pipeline.go:95]: Trick question, it never fails. Good catch

File: pkg/gstreamer/bin.go [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+			}
+		}
+	} else {
+		// link src bins to sink bins
+		srcs := getPeerSrcs(b.srcs)
+		sinks := getPeerSinks(b.sinks)
+
+		addQueues := len(sinks) > 1
+		for _, src := range srcs {
+			src.mu.Lock()
```
[the remark this answers is not in the document]
Comment (davidzhao) [reply pkg/gstreamer/bin.go:319]: right, I just noticed it because both locks were held. but I guess it's covered by the main Bin lock

**Review by davidzhao — APPROVED** [review]
🚀 🚀

**Pull request merged**